scvi/dataset/dataset.py

```diff
@@ -1468,6 +1476,105 @@ def get_batch_mask_cell_measurement(self, attribute_name: str):
+            of enrichment of zeros for each gene. Default: ``10000``.
+        :param silent: If ``True``, disables the progress bar. Default ``False``.
+        :param batch_size: Size of temporary matrix for incremental calculation. Larger
+            is faster but requires more RAM or GPU memory. (The default should be fine unless
+            there are hundreds of millions cells or millions of genes.) Default ``5000``.
+
+        :return: A ``pd.DataFrame`` with per-gene statistics.
+
+    """
+    from tqdm import tqdm
+
+    use_cuda = use_cuda and torch.cuda.is_available()
+    dev = torch.device('cuda:0' if use_cuda else 'cpu')
+
+    # Make sure were are working with a sparse CSC matrix.
+    from scipy import sparse
+    X = sparse.csc_matrix(X)
+
+    # Calculate empirical statistics.
+    scaled_means = torch.from_numpy(X.sum(0) / X.sum())[0].to(dev)
+    total_counts = torch.from_numpy(X.sum(1))[:, 0].to(dev)
+
+    observed_fraction_zeros = torch.from_numpy(1. - (X > 0).sum(0) / X.shape[0])[0].to(dev)
+
+    # Calculate probability of zero for a Poisson model.
+    # Perform in batches to save memory.
+    n_batches = total_counts.shape[0] // batch_size
+
+    expected_fraction_zeros = torch.zeros(scaled_means.shape).to(dev)
+
+    for i in range(n_batches):
+        total_counts_batch = total_counts[i * batch_size:(i + 1) * batch_size]
+        # Use einsum for outer product.
+        expected_fraction_zeros += torch.exp(-torch.einsum('i,j->ij', [scaled_means, total_counts_batch])).sum(1)
+
+    total_counts_batch = total_counts[(i + 1) * batch_size:]
+    expected_fraction_zeros += torch.exp(-torch.einsum('i,j->ij', [scaled_means, total_counts_batch])).sum(1)    
+    expected_fraction_zeros /= X.shape[0]
+
+    # Compute probability of enriched zeros through sampling from Binomial distributions.
+    observed_zero = torch.distributions.Binomial(probs=observed_fraction_zeros)
+    expected_zero = torch.distributions.Binomial(probs=expected_fraction_zeros)
+
+    extra_zeros = torch.zeros(expected_fraction_zeros.shape).to(dev)
+    for i in tqdm(range(n_samples), disable=silent, desc='Sampling from binomial'):
+        extra_zeros += (observed_zero.sample() > expected_zero.sample())
+
+    prob_zero_enrichment = (extra_zeros / n_samples).cpu().numpy()
+
+    obs_frac_zeros = observed_fraction_zeros.cpu().numpy()
+    exp_frac_zeros = expected_fraction_zeros.cpu().numpy()
+
+    # Clean up memory (tensors seem to stay in GPU unless actively deleted).
+    del scaled_means
+    del total_counts
+    del expected_fraction_zeros
+    del observed_fraction_zeros
+    del extra_zeros
+
+    if use_cuda:
+        torch.cuda.empty_cache()
+
+    # Create result DataFrame
+
+    df = pd.DataFrame({
+        'observed_fraction_zeros': obs_frac_zeros,
+        'expected_fraction_zeros': exp_frac_zeros,
+        'prob_zero_enrichment': prob_zero_enrichment,
+        'highly_variable': False
+    })
```

Comment:
My other general thought is that it would be great if we could have this run per batch (RNA batch not gpu batch) and then sort by how many batches it’s highly variable followed by the prob zero enrichment. You can follow what is done in the seurat v3 code.

Reply:
I was thinking about this too when I looked at the tests. I'll give it a try and see how it goes.